File: tw-gradcs-linear-algebra/04-orthogonality-and-least-squares/04-gram-schmidt-process/c/gram_schmidt.c
/**
 * Gram-Schmidt 正交化 (Gram-Schmidt Process)
 *
 * 編譯：gcc -std=c99 -O2 gram_schmidt.c -o gram_schmidt -lm
 * 執行：./gram_schmidt
 */

#include <stdio.h>
#include <math.h>

#define MAX_DIM 10

void print_separator(const char* title) {
    printf("\n");
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n%s\n", title);
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n");
}

void print_vector(const char* name, const double* v, int n) {
    printf("%s = [", name);
    for (int i = 0; i < n; i++) {
        printf("%.4f", v[i]);
        if (i < n - 1) printf(", ");
    }
    printf("]\n");
}

double dot_product(const double* x, const double* y, int n) {
    double result = 0.0;
    for (int i = 0; i < n; i++) result += x[i] * y[i];
    return result;
}

double vector_norm(const double* x, int n) {
    return sqrt(dot_product(x, x, n));
}

void modified_gram_schmidt(double A[MAX_DIM][MAX_DIM], double Q[MAX_DIM][MAX_DIM], int num_vecs, int dim) {
    // 複製
    for (int i = 0; i < num_vecs; i++)
        for (int j = 0; j < dim; j++)
            Q[i][j] = A[i][j];

    for (int j = 0; j < num_vecs; j++) {
        for (int i = 0; i < j; i++) {
            double coeff = dot_product(Q[i], Q[j], dim) / dot_product(Q[i], Q[i], dim);
            for (int k = 0; k < dim; k++)
                Q[j][k] -= coeff * Q[i][k];
        }
    }
}

void normalize_vectors(double Q[MAX_DIM][MAX_DIM], int num_vecs, int dim) {
    for (int i = 0; i < num_vecs; i++) {
        double norm = vector_norm(Q[i], dim);
        for (int j = 0; j < dim; j++)
            Q[i][j] /= norm;
    }
}

int verify_orthogonality(double Q[MAX_DIM][MAX_DIM], int num_vecs, int dim) {
    for (int i = 0; i < num_vecs; i++) {
        for (int j = i + 1; j < num_vecs; j++) {
            if (fabs(dot_product(Q[i], Q[j], dim)) > 1e-10)
                return 0;
        }
    }
    return 1;
}

int main() {
    print_separator("Gram-Schmidt 正交化示範 (C)");

    double A[MAX_DIM][MAX_DIM] = {
        {1.0, 1.0, 0.0},
        {1.0, 0.0, 1.0},
        {0.0, 1.0, 1.0}
    };
    int num_vecs = 3, dim = 3;

    printf("輸入向量組：\n");
    for (int i = 0; i < num_vecs; i++) {
        char name[10];
        sprintf(name, "a%d", i+1);
        print_vector(name, A[i], dim);
    }

    double Q[MAX_DIM][MAX_DIM];
    modified_gram_schmidt(A, Q, num_vecs, dim);

    printf("\n正交化結果（MGS）：\n");
    for (int i = 0; i < num_vecs; i++) {
        char name[10];
        sprintf(name, "q%d", i+1);
        print_vector(name, Q[i], dim);
        printf("    ‖q%d‖ = %.4f\n", i+1, vector_norm(Q[i], dim));
    }

    printf("\n正交？ %s\n", verify_orthogonality(Q, num_vecs, dim) ? "true" : "false");

    printf("\n內積驗證：\n");
    printf("q₁ · q₂ = %.6f\n", dot_product(Q[0], Q[1], dim));
    printf("q₁ · q₃ = %.6f\n", dot_product(Q[0], Q[2], dim));
    printf("q₂ · q₃ = %.6f\n", dot_product(Q[1], Q[2], dim));

    print_separator("標準正交化");

    normalize_vectors(Q, num_vecs, dim);

    printf("標準正交向量組：\n");
    for (int i = 0; i < num_vecs; i++) {
        char name[10];
        sprintf(name, "e%d", i+1);
        print_vector(name, Q[i], dim);
        printf("    ‖e%d‖ = %.4f\n", i+1, vector_norm(Q[i], dim));
    }

    print_separator("總結");
    printf("\nGram-Schmidt 核心公式：\n\n");
    printf("proj_q(a) = (qᵀa / qᵀq) q\n\n");
    printf("q₁ = a₁\n");
    printf("qₖ = aₖ - Σᵢ proj_{qᵢ}(aₖ)\n\n");
    printf("eᵢ = qᵢ / ‖qᵢ‖\n\n");

    for (int i = 0; i < 60; i++) printf("=");
    printf("\n示範完成！\n");
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n");

    return 0;
}

File: tw-gradcs-linear-algebra/04-orthogonality-and-least-squares/04-gram-schmidt-process/javascript/gram_schmidt.js
/**
 * Gram-Schmidt 正交化 (Gram-Schmidt Process)
 *
 * 執行：node gram_schmidt.js
 */

function printSeparator(title) {
    console.log();
    console.log('='.repeat(60));
    console.log(title);
    console.log('='.repeat(60));
}

function formatNumber(n) {
    return n.toFixed(4);
}

function printVector(name, v) {
    const formatted = v.map(x => formatNumber(x)).join(', ');
    console.log(`${name} = [${formatted}]`);
}

function dotProduct(x, y) {
    return x.reduce((sum, xi, i) => sum + xi * y[i], 0);
}

function vectorNorm(x) {
    return Math.sqrt(dotProduct(x, x));
}

function scalarMultiply(c, x) {
    return x.map(xi => c * xi);
}

function vectorSubtract(x, y) {
    return x.map((xi, i) => xi - y[i]);
}

function normalize(x) {
    return scalarMultiply(1 / vectorNorm(x), x);
}

/**
 * Modified Gram-Schmidt
 */
function modifiedGramSchmidt(A) {
    const Q = A.map(row => [...row]);

    for (let j = 0; j < Q.length; j++) {
        for (let i = 0; i < j; i++) {
            const coeff = dotProduct(Q[i], Q[j]) / dotProduct(Q[i], Q[i]);
            Q[j] = vectorSubtract(Q[j], scalarMultiply(coeff, Q[i]));
        }
    }

    return Q;
}

function gramSchmidtNormalized(A) {
    return modifiedGramSchmidt(A).map(q => normalize(q));
}

function verifyOrthogonality(Q) {
    for (let i = 0; i < Q.length; i++) {
        for (let j = i + 1; j < Q.length; j++) {
            if (Math.abs(dotProduct(Q[i], Q[j])) > 1e-10) return false;
        }
    }
    return true;
}

function main() {
    printSeparator('Gram-Schmidt 正交化示範 (JavaScript)');

    const A = [
        [1, 1, 0],
        [1, 0, 1],
        [0, 1, 1]
    ];

    console.log('輸入向量組：');
    A.forEach((a, i) => printVector(`a${i+1}`, a));

    const Q = modifiedGramSchmidt(A);

    console.log('\n正交化結果（MGS）：');
    Q.forEach((q, i) => {
        printVector(`q${i+1}`, q);
        console.log(`    ‖q${i+1}‖ = ${formatNumber(vectorNorm(q))}`);
    });

    console.log(`\n正交？ ${verifyOrthogonality(Q)}`);

    console.log('\n內積驗證：');
    console.log(`q₁ · q₂ = ${dotProduct(Q[0], Q[1]).toFixed(6)}`);
    console.log(`q₁ · q₃ = ${dotProduct(Q[0], Q[2]).toFixed(6)}`);
    console.log(`q₂ · q₃ = ${dotProduct(Q[1], Q[2]).toFixed(6)}`);

    printSeparator('標準正交化');

    const E = gramSchmidtNormalized(A);

    console.log('標準正交向量組：');
    E.forEach((e, i) => {
        printVector(`e${i+1}`, e);
        console.log(`    ‖e${i+1}‖ = ${formatNumber(vectorNorm(e))}`);
    });

    printSeparator('總結');
    console.log(`
Gram-Schmidt 核心公式：

proj_q(a) = (qᵀa / qᵀq) q

q₁ = a₁
qₖ = aₖ - Σᵢ proj_{qᵢ}(aₖ)

eᵢ = qᵢ / ‖qᵢ‖
`);

    console.log('='.repeat(60));
    console.log('示範完成！');
    console.log('='.repeat(60));
}

main();
